### PR TITLE
added basic integration test setup

### PR DESCRIPTION
**Pull request by xh3b4sd**
```
$ go test -tags=integration ./informer
ok  	github.com/giantswarm/operatorkit/informer	24.155s
```

### CODE REVIEW

File: crd/crd.go [hunk 1]
```diff
@@ -88,6 +88,10 @@ func New(config Config) (*CRD, error) {
 	return newCRD, nil
 }
 
+func (c *CRD) CreateEndpoint() string {
+	return filepath.Join("apis", c.Group(), c.Version(), c.Plural())
+}
```
Comment (xh3b4sd) [inline crd/crd.go]: In the tests I noticed it makes sense to have a dedicated method to get the endpoint which we can use to create custom objects. Sneaking that in here.

File: crd/fake/crd.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package fake
+
+import (
+	"fmt"
+
+	"github.com/giantswarm/operatorkit/crd"
+)
+
+func MustNewCRD() *crd.CRD {
```
Comment (xh3b4sd) [inline crd/fake/crd.go]: All the fake CRD magic is for testing. It helps with the CRD/CRO details that we want to ensure in our tests.

File: crd/fake/cro.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		},
+		Spec: Spec{
+			ID: ID,
+		},
+	}
+
+	b, err := json.Marshal(newCustomObject)
+	if err != nil {
+		panic(fmt.Sprintf("%#v", err))
+	}
```
Comment (xh3b4sd) [inline crd/fake/cro.go]: This is a bit odd. We use a rest client to create/delete/update custom objects. With 1.8 I am not able to get the rest client encoder working to accept a custom object as `runtime.Object`. That is why I just use the json library to marshal it.

File: informer/informer.go [hunk 4]
```diff
@@ -5,7 +5,7 @@
 //     - The informer is able to watch all kinds of objects as soon as a proper
 //       watch endpoint and a factory implementing ZeroObjectFactory is given.
 //     - Events for objects that are created, deleted or updated are dispatched
-//       immeditiatelly.
+//       immediately.
```
Comment (xh3b4sd) [inline informer/informer.go]: Boyscouting.

File: informer/informer.go [hunk 5]
```diff
@@ -41,8 +41,8 @@ const (
 // Config represents the configuration used to create a new Informer.
 type Config struct {
 	// Dependencies.
-	BackOff    backoff.BackOff
-	RestClient rest.Interface
+	BackOff        backoff.BackOff
+	WatcherFactory WatcherFactory
```
Comment (xh3b4sd) [inline informer/informer.go]: I eased the configuration and execution of the watch using a watcher factory. This was done in the first place when I tried to write useful unit tests. I couldn't find a way to do this and I ended up with the integration test introduced in this PR. I kept the interface/configuration improvements though.

File: informer/informer.go [hunk 6]
```diff
@@ -145,7 +145,12 @@ func (i *Informer) Watch(ctx context.Context, endpoint string, factory ZeroObjec
 		for {
 			select {
 			case <-done:
-			case event := <-eventChan:
+				return
```
Comment (xh3b4sd) [inline informer/informer.go]: This fixes a bug catched by the integration tests.

File: informer/informer.go [hunk 7]
```diff
@@ -145,7 +145,12 @@ func (i *Informer) Watch(ctx context.Context, endpoint string, factory ZeroObjec
...
+			case event, ok := <-eventChan:
+				if !ok {
+					return
+				}
```
Comment (xh3b4sd) [inline informer/informer.go]: This fixes a bug catched by the integration tests.

File: informer/informer.go [hunk 8]
```diff
@@ -224,18 +229,29 @@ func (i *Informer) Watch(ctx context.Context, endpoint string, factory ZeroObjec
...
+	m, err := meta.Accessor(event.Object)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	t := m.GetDeletionTimestamp()
+	if t == nil {
+		updateChan <- event
+	} else {
+		deleteChan <- event
+	}
```
Comment (xh3b4sd) [inline informer/informer.go]: This is a result of testing finalizers with CRDs on 1.8. When a event object is deleted and it has a finalizer, it is replayed on resync period and then send to the delete channel.

File: informer/informer_integration_basic_test.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+import (
+	"context"
+	"testing"
+	"time"
+)
+
+// Test_Informer_Integration_Basic is a integration test for basic cache
+// informer operations. The test verifies the informer is operating as expected
+// when processing basic sequences of creating and deleting runtime objects.
+func Test_Informer_Integration_Basic(t *testing.T) {
```
Comment (xh3b4sd) [inline informer/informer_integration_basic_test.go]: This is the actual integration test.

File: informer/informer_integration_test.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
+// +build integration
+
+package informer
```
Comment (xh3b4sd) [inline informer/informer_integration_test.go]: This is the integration test setup. Stolen from tprstorage.

File: informer/informer_integration_test.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+		homePath := func(relativePath string) string {
+			if err != nil {
+				return ""
+			}
+			return path.Join(u.HomeDir, relativePath)
+		}
+
+		var serverDefault string
+		{
+			out, err := exec.Command("minikube", "ip").Output()
```
Comment (fgimenez) [inline informer/informer_integration_test.go]: With `shipyard` you could spin up a remote minikube in aws and run these tests on circleCI, let me know if i can help with that :)

Comment (JosephSalisbury) [reply informer/informer_integration_test.go]: I want this too :D

Comment (xh3b4sd) [reply informer/informer_integration_test.go]: 100%, I pinged you to get feedback on this. I would like to get the PR done first and integrate it into Circle separately. Just let me know how this works.

File: crd/crd.go [hunk 12]
```diff
@@ -88,6 +88,10 @@ func New(config Config) (*CRD, error) {
 	return newCRD, nil
 }
 
+func (c *CRD) CreateEndpoint() string {
+	return filepath.Join("apis", c.Group(), c.Version(), c.Plural())
```
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply crd/crd.go]: Done.

File: informer/informer_integration_basic_test.go [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+	// We define a general control goroutine to stop test execution on errors or
+	// after timeouts. The timeout is 25 seconds because of two resync periods,
+	// one rate limit wait period, plus some time buffer for all the test magic
+	// happening here: (2 * 10) + 2 + buffer.
+	go func() {
+		for {
+			select {
+			case <-time.After(25 * time.Second):
+				t.Fatalf("expected %#v got %#v", "proper test execution", "timeout")
```
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply informer/informer_integration_basic_test.go]: Done mofo.

File: informer/informer_integration_basic_test.go [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+		}
+	}()
+
+	// This is the first time we want to catch an event from the informer watch.
+	// This should give us the runtime object we created before starting the
+	// watch.
+	{
+		select {
+		case <-deleteChan:
+			t.Fatalf("expected %#v got %#v", "update event", "delete event")
```
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply informer/informer_integration_basic_test.go]: Done mofo.

File: informer/informer_integration_basic_test.go [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+		case e := <-updateChan:
+			testAssertCROWithID(t, e, "al7qy")
+		}
+	}
+
+	// We create another runtime object. This should be received immediately.
+	{
+		testCreateCRO(t, "al8qy")
```
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply informer/informer_integration_basic_test.go]: Done.

File: informer/informer_integration_basic_test.go [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+
+	// This is the first time we want to catch an event from the informer watch.
+	// This should give us the runtime object we created before starting the
+	// watch.
+	{
+		select {
+		case <-deleteChan:
+			t.Fatalf("expected %#v got %#v", "update event", "delete event")
+		case e := <-updateChan:
+			testAssertCROWithID(t, e, "al7qy")
```
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply informer/informer_integration_basic_test.go]: Done.

**Comment by xh3b4sd** [comment]
Fishing for approvals. Please have another look.

**Comment by fgimenez** [comment]
> I want this too :D

we still need https://github.com/giantswarm/shipyard/pull/4 merged first, comments welcome! :)

**Pull request merged**